### PR TITLE
SONARJAVA-1314 Locks left unlocked

### PR DESCRIPTION
**Pull request by DidierBesset**

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+          track(expression, States.LOCKED);
+        } else if (TRY_LOCK_METHOD_NAME.equals(methodName) && !isDirectedAtField(expression)) {
+          transitionFactory = new TransitionFactory(expression, programState.peekValue());
+        } else if (UNLOCK_METHOD_NAME.equals(methodName)) {
+          track(expression, States.UNLOCKED);
+        }
+      }
+    }
+
+    private boolean isDirectedAtField(ExpressionTree expression) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java]: MAJOR Make "isDirectedAtField" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  priority = Priority.CRITICAL,
+  tags = {"bug", "cwe", "multi-threading"})
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.SYNCHRONIZATION_RELIABILITY)
+@SqaleConstantRemediation("20min")
+public class LocksNotUnlockedCheck extends SECheck implements JavaFileScanner {
+
+  private static final Logger LOG = LoggerFactory.getLogger(LocksNotUnlockedCheck.class);
+
+  private static enum States {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java]: MAJOR Remove this redundant "static" qualifier. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2786)

File: java-squid/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java [hunk 3]
```diff
@@ -141,12 +118,14 @@ public PostStatementVisitor(CheckerContext context) {
     public void visitNewClass(NewClassTree syntaxNode) {
       if (isOpeningResource(syntaxNode)) {
         final SymbolicValue instanceValue = programState.peekValue();
-        programState = programState.addConstraint(instanceValue, NullConstraint.OPENED);
+        if (instanceValue.wrappedValue() == instanceValue) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

File: java-squid/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java [hunk 4]
```diff
@@ -62,76 +60,55 @@
...
-    @Override
-    protected void scan(List<? extends Tree> trees) {
-      // Cut recursive processing
-    }
-
-    @Override
-    protected void scan(ListTree<? extends Tree> listTree) {
-      // Cut recursive processing
-    }
+  private static enum States {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: MAJOR Remove this redundant "static" qualifier. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2786)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 4 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Add a new line at the end of this file.](https://github.com/SonarSource/sonar-java/blob/8a2de92bdd8fc1d4c1057e307d54b5dc7250e6a2/java-squid/src/main/java/org/sonar/java/se/checks/CheckerTreeNodeVisitor.java) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00113)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 4 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Add a new line at the end of this file.](https://github.com/SonarSource/sonar-java/blob/8a2de92bdd8fc1d4c1057e307d54b5dc7250e6a2/java-squid/src/main/java/org/sonar/java/se/checks/CheckerTreeNodeVisitor.java) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00113)

File: java-checks/src/main/java/org/sonar/java/checks/CheckList.java [hunk 5]
```diff
@@ -21,6 +21,7 @@
 
 import com.google.common.collect.ImmutableList;
 import org.sonar.java.se.checks.ConditionAlwaysTrueOrFalseCheck;
+import org.sonar.java.se.checks.LocksNotUnlockedCheck;
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/CheckList.java]: Isn't this a unused import ?

File: java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java (hunk 1 above)
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java]: What does DirectedAt  means ?

File: java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    private static final String TRY_LOCK_METHOD_NAME = "tryLock";
+    private static final String UNLOCK_METHOD_NAME = "unlock";
+
+    private TransitionFactory transitionFactory = null;
+
+    public PreStatementVisitor(CheckerContext context) {
+      super(context.getState());
+    }
+
+    TransitionFactory transitionFactory() {
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java]: As you are working in a private static inner class I am really unsure about the pertinence of a getter in this case.

File: java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+  }
+
+  private static class PreStatementVisitor extends CheckerTreeNodeVisitor {
+
+    private static final String LOCK = "java.util.concurrent.locks.Lock";
+    private static final String LOCK_METHOD_NAME = "lock";
+    private static final String TRY_LOCK_METHOD_NAME = "tryLock";
+    private static final String UNLOCK_METHOD_NAME = "unlock";
+
+    private TransitionFactory transitionFactory = null;
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java]: Detail but no need for initializer.

File: java-squid/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java (hunk 3 above)
Comment (DidierBesset) [inline java-squid/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: Here, it is important to check for the same object and not equality. How to overwrite this?

File: java-squid/src/test/files/se/InsidePercentile.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+import com.sonarsource.quipu.internal.data.Measurement;
+import com.sonarsource.quipu.data.QuipuResult;
+import com.sonarsource.quipu.unit.Quantity;
+ 
+import java.util.Formatter;
+import java.util.List;
+ 
+import static com.sonarsource.quipu.internal.statistics.Utils.*;
+ 
+public class InsidePercentile implements StatisticsTest {
```
Comment (benzonico) [inline java-squid/src/test/files/se/InsidePercentile.java]: What does this file has to do with anything ?

File: java-squid/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java (hunk 3 above)
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: Fine by precising with a comment why we should keep it.

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Add a new line at the end of this file.](https://github.com/SonarSource/sonar-java/blob/a43d82bb6325dd0dadaa530e909ec708e6bf75a2/java-squid/src/main/java/org/sonar/java/se/checks/CheckerTreeNodeVisitor.java) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00113)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Add a new line at the end of this file.](https://github.com/SonarSource/sonar-java/blob/a43d82bb6325dd0dadaa530e909ec708e6bf75a2/java-squid/src/main/java/org/sonar/java/se/checks/CheckerTreeNodeVisitor.java) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00113)

File: java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+  priority = Priority.CRITICAL,
+  tags = {"bug", "cwe", "multi-threading"})
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.SYNCHRONIZATION_RELIABILITY)
+@SqaleConstantRemediation("20min")
+public class LocksNotUnlockedCheck extends SECheck implements JavaFileScanner {
+
+  private static final Logger LOG = LoggerFactory.getLogger(LocksNotUnlockedCheck.class);
+
+  private enum States {
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java]: Naming thing States in this context is rather unfortunate as we already have the notion of ProgramState which is precise and fairly different. What about status ? or something more specific ?

File: java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+    LOCKED, UNLOCKED;
+  }
+
+  private static class TransitionFactory {
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java]: I am really unsure of the purpose of this class... A factory is producing something: here it adds constraints to a program state.

I also struggle to determine what is the purpose of setting it to null in between pre and post check in statements.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Add a new line at the end of this file.](https://github.com/SonarSource/sonar-java/blob/b09a1c37b3ad42b512a699b03592ec9bcfb3b209/java-squid/src/main/java/org/sonar/java/se/checks/CheckerTreeNodeVisitor.java) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00113)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Add a new line at the end of this file.](https://github.com/SonarSource/sonar-java/blob/22e73c52ee78bca57d0a3930e826e4b01ed386aa/java-squid/src/main/java/org/sonar/java/se/checks/CheckerTreeNodeVisitor.java) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00113)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Add a new line at the end of this file.](https://github.com/SonarSource/sonar-java/blob/22e73c52ee78bca57d0a3930e826e4b01ed386aa/java-squid/src/main/java/org/sonar/java/se/checks/CheckerTreeNodeVisitor.java) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00113)

File: java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+    TransitionFactory transitionFactory;
+
+    public PreStatementVisitor(CheckerContext context) {
+      super(context.getState());
+    }
+
+    @Override
+    public void visitMemberSelectExpression(MemberSelectExpressionTree syntaxNode) {
+      final ExpressionTree expression = syntaxNode.expression();
+      if (expression.symbolType().isSubtypeOf(LOCK)) {
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java]: Please use MethodInvocationMatcher

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Add a new line at the end of this file.](https://github.com/SonarSource/sonar-java/blob/22e73c52ee78bca57d0a3930e826e4b01ed386aa/java-squid/src/main/java/org/sonar/java/se/checks/CheckerTreeNodeVisitor.java) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00113)

File: java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java (hunk 11 above)
Comment (DidierBesset) [inline java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java]: Of course, but this class is located in java-checks which depends on java-squid. Thus, it cannot be made accessible unless you want to jumble the whole project structure...

File: java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java (hunk 10 above)
Comment (DidierBesset) [inline java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java]: This class is the factory for lock/unlock constraints.
Would it help to name it ConstraintFactory? ConstraintGenerator?
Pick up a name...

As for the null setting, it is just that I want to ensure it is reset in any possible situation. Strictly speaking, the only one needed is the one at the end of the post statement. So what?

File: java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java (hunk 11 above)
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java]: fair enough !

File: java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.SYNCHRONIZATION_RELIABILITY)
+@SqaleConstantRemediation("20min")
+public class LocksNotUnlockedCheck extends SECheck implements JavaFileScanner {
+
+  private enum Status {
+    LOCKED, UNLOCKED;
+  }
+
+  private static class TryLockSymbolicValue extends SymbolicValue {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java]: MAJOR Override this superclass' "equals" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2160)

File: java-squid/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java [hunk 13]
```diff
@@ -62,91 +61,101 @@
...
         }
       }
-      return isCloseable(type);
     }
+  }
 
-    protected static boolean isCloseable(Type type) {
-      return type.isSubtypeOf(JAVA_IO_CLOSEABLE);
-    }
+  public static class ResourceWrapperSymbolicValue extends SymbolicValue {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: MAJOR Override this superclass' "equals" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2160)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: java-squid/src/main/java/org/sonar/java/se/ConstraintManager.java [hunk 14]
```diff
@@ -27,10 +27,13 @@
 public class ConstraintManager {
 
   private int counter = ProgramState.EMPTY_STATE.constraintsSize();
-  private SymbolicValue wrappedValue;
+  private SymbolicValueFactory symbolicValueFactory;
 
-  public void setWrappedValue(SymbolicValue wrappedValue) {
-    this.wrappedValue = wrappedValue;
+  public void setValueFactory(SymbolicValueFactory valueFactory) {
+    if (symbolicValueFactory != null) {
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/se/ConstraintManager.java]: You can use guava preconditions check in this case : https://google-collections.googlecode.com/svn/trunk/javadoc/com/google/common/base/Preconditions.html#checkState(boolean) 

Simplifies a bit the code.

File: java-squid/src/main/java/org/sonar/java/se/SymbolicValueFactory.java [hunk 15]
```diff
@@ -17,13 +17,11 @@
...
-import org.junit.runners.Suite;
-import org.sonar.java.checks.LocksNotUnlockedCheckTest;
+import org.sonar.plugins.java.api.tree.Tree;
 
-@RunWith(Suite.class)
-@Suite.SuiteClasses({LocksNotUnlockedCheckTest.class})
-public class SymExecChecks {
+public interface SymbolicValueFactory {
+
+  SymbolicValue createSymbolicValue(int counter, Tree syntaxNode);
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/se/SymbolicValueFactory.java]: In the interface the int parameter should be probably called id. It is only a counter in constraint manager for convenience.

File: java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+        }
+      }
+    }
+
+    private static boolean isMemberSelectActingOnField(IdentifierTree expression) {
+      final Symbol symbol = expression.symbol();
+      return ProgramState.isField(symbol);
+    }
+
+    private void track(ExpressionTree syntaxNode, SymbolicValue instanceValue, Status status) {
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java]: why `track` ? why not addConstraint ? and thus ordering parameter with symbolicValue parameter (that should be called like this) first.

File: java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+    private final ExpressionTree expression;
+    private final SymbolicValue operand;
+
+    TryLockSymbolicValueFactory(ExpressionTree expression, final SymbolicValue operand) {
+      this.expression = expression;
+      this.operand = operand;
+    }
+
+    @Override
+    public SymbolicValue createSymbolicValue(int counter, Tree syntaxNode) {
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/se/checks/LocksNotUnlockedCheck.java]: syntaxNode is unused here.

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: java-squid/src/test/files/se/UnclosedResourcesCheck.java [hunk 18]
```diff
@@ -240,6 +248,18 @@ public void correctHandlingOfJarFile(String fileName) {
       closeJar(jar);
     }
   }
+
+  public void getDirectivesFromFile(File aFile) {
+    BufferedReader reader = null;
+    try {
+      reader = new BufferedReader(new FileReader(aFile));
+      reader.read();
+//    } catch(IOException e) { // Causes a problem when decommented!
```
Comment (benzonico) [inline java-squid/src/test/files/se/UnclosedResourcesCheck.java]: which one ?

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: java-squid/src/test/files/se/UnclosedResourcesCheck.java (hunk 18 above)
Comment (DidierBesset) [inline java-squid/src/test/files/se/UnclosedResourcesCheck.java]: What do you mean? This is just test code copied verbatim from a file from "ruling"...

Comment (benzonico) [inline java-squid/src/test/files/se/UnclosedResourcesCheck.java]: So then please remove the comment, I can't know (and neitheir further readers) that you are not the one who wrote this... example should not be taken in extenso in unit test files but stripped down to cases.

**Pull request merged**